articles/heap-data-structure-python/index.md

```diff
@@ -0,0 +1,202 @@
+
+#### Maxheap using List
+
+We are going to do the list implementation of a heap. In this, the heap's level-order traversal would be stored as an array/list.
+
+![Level Order Traversal](/engineering-education/heap-data-structure-python/level-order-traversal.png)
+
+_Figure: Level-Order Traversal ([Source](https://www.quora.com/What-is-level-order-traversal-in-a-binary-tree))_
+
+_Note - Level-Order Traversal is a recursive traversal where the root is processed first, followed by the children of the root. This is followed by the grandchildren of the root till all the nodes are processed._
+
+In the array representation of a heap, for an element in array index i,
+
+- The Parent Node would be at position floor((i-1)/2).
+- The Left Child would be at position 2\*i + 1.
+- The Right Child would be at position 2\*i + 2.
+
+Let us first define the Heap class.
+
+```python
+class MaxHeap:
+    def __init__(self):
+        self.heap = []
+```
+
+This initiates a heap as a list. Now, let us define our methods.
+
+```python
+class MaxHeap:
+    def __init__(self):
+        # Initialize a heap using list
+        self.heap = []
+
+    def getParentPosition(self, i):
+        # The parent is located at floor((i-1)/2)
+        return int((i-1)/2)
+
+    def getLeftChildPosition(self, i):
+        # The left child is located at 2 * i + 1
+        return 2*i+1
+
+    def getRightChildPosition(self, i):
+        # The right child is located at 2 * i + 2
+        return 2*i+2
+
+    def hasParent(self, i):
+        # This function checks if the given node has a parent or not
+        return self.getParentPosition(i) < len(self.heap)
+
+    def hasLeftChild(self, i):
+        # This function checks if the given node has a left child or not
+        return self.getLeftChildPosition(i) < len(self.heap)
+
+    def hasRightChild(self, i):
+        # This function checks if the given node has a right child or not
+        return self.getRightChildPosition(i) < len(self.heap)
+
+    def insert(self, key):
+        self.heap.append(key) # Adds the key to the end of the list
+        self.heapify(len(self.heap) - 1) # Re-arranges the heap to maintain the heap property
+
+    def getMax(self):
+        return self.heap[0] # Returns the largest value in the heap in O(1) time.
+
+    def heapify(self, i):
+        while(self.hasParent(i) and self.heap[i] > self.heap[self.getParentPosition(i)]): # Loops until it reaches a leaf node
+            self.heap[i], self.heap[self.getParentPosition(i)] = self.heap[self.getParentPosition(i)], self.heap[i] # Swap the values
+            i = self.getParentPosition(i) # Resets the new position
+
+    def printHeap(self):
+        print(self.heap) # Prints the heap
+```
+
+#### Minheap using Heapq
+
+We have successfully implemented a heap using a list. Now, let us use the **`heapq`** library in Python to implement a minheap.
+
+```python
+import heapq
+class Minheap:
+    def __init__(self, minheap):
```

Comment:
It's not clear what the reader should be passing into this constructor. What is `minheap`?

Reply:
I've added a code comment that specifies what minheap is.
It's a list that we pass as input. the list that we wanna convert to a heap.